dynawo/sources/Models/CPP/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp

```diff
@@ -0,0 +1,579 @@
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::U2() const {
+    return Ur() * Ur() + Ui() * Ui();
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::P() const {
+    return yLocal_[PYNum_];
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::Q() const {
+    return yLocal_[QYNum_];
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalF(double /*t*/, propertyF_t type) {
+    if (getIsInitProcess())
+      return;
+    if (isConnected()) {
+      double UfRawValue = UfRaw();
+      if (UMaxPuReached_ == true) {
+        UfRawValue = UMaxPu_;
+      }
+      if (UMinPuReached_ == true) {
+        UfRawValue = UMinPu_;
+      }
+      if (type == ALGEBRAIC_EQ) {
+        fLocal_[1] = Uf() - UfRawValue;
+        fLocal_[2] = Ir() - (P() * Ur() + Q() * Ui()) / (U() * U());
+        fLocal_[3] = Ii() - (P() * Ui() - Q() * Ur()) / (U() * U());
+        fLocal_[4] = P() - P0_ * pow(U() / Uf(), alpha_);
+        fLocal_[5] = Q() - Q0_ * pow(U() / Uf(), beta_);
+      } else if (type == DIFFERENTIAL_EQ) {
+        fLocal_[0] = Tf_ * UfRawPrim() - (U() - UfRaw());
+      } else {
+        fLocal_[0] = Tf_ * UfRawPrim() - (U() - UfRaw());
+        fLocal_[1] = Uf() - UfRawValue;
+        fLocal_[2] = Ir() - (P() * Ur() + Q() * Ui()) / (U() * U());
+        fLocal_[3] = Ii() - (P() * Ui() - Q() * Ur()) / (U() * U());
+        fLocal_[4] = P() - P0_ * pow(U() / Uf(), alpha_);
+        fLocal_[5] = Q() - Q0_ * pow(U() / Uf(), beta_);
+      }
+    } else {
+      fLocal_[0] = UfRaw();
+      fLocal_[1] = Uf();
+      fLocal_[2] = Ir();
+      fLocal_[3] = Ii();
+      fLocal_[4] = P();
+      fLocal_[5] = Q();
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::setFequations() {
+    if (getIsInitProcess())
+      return;
+    if (isConnected()) {
+      unsigned int index = 0;
+      fEquationIndex_[index] = std::string("Tf_*UfPrim() - (U() - Uf()):").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("Uf() - UfRaw():").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("Ir() - (P() * Ur() + Q() * Ui()) / (U() * U()): ").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("Ii() - (P() * Ui() - Q() * Ur()) / (U() * U()): ").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("P() - P0_ * (U()/Uf())^alpha_ :").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("Q() - P0_ * (U()/Uf())^beta_ :").append(staticId());
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalZ(const double & /*t*/) {
+    State connectionState1 = CLOSED;
+    if (zLocal_[0] > 0) {
+      connectionState1 = OPEN;
+    }
+    State connectionState2 = CLOSED;
+    if (zLocal_[1] > 0) {
+      connectionState2 = OPEN;
+    }
+    if (connectionState1 != connectionState1_prev) {
+      setConnected(connectionState1);
+      connectionState1_prev = connectionState1;
+    }
+    if (connectionState2 != connectionState2_prev) {
```

Comment:
if we have state1 = CLOSED and state2 = OPEN and then at the same timestep state1 = OPEN and state2 = CLOSED then the connection state will be CLOSED whereas it should be opened